### ISSUE TITLE
关于FM和KGCN训练的问题

### ISSUE THREAD

**Issue by MGzhou** [RUCAIBox/RecBole#958]
## 首先感谢你们做出这么棒的库

我最近在是用库时遇到了一些疑问，很希望得到你们的解答。谢谢。

## 1 针对FM训练的问题

训练代码如下（默认模式）：

```python
from recbole.quick_start import run_recbole
run_recbole(model='FM', dataset='ml-100k')
```

他的其中一个配置是`split_ratio = [0.8, 0.1, 0.1]`  对应【train data训练集, valid data验证集 and test data测试集】

`eval_batch_size = 4096`

`train_batch_size = 2048`

先看我的第一个批次的输出

![image-20210907192331372](https://gitee.com/MGzhou/zhoupic/raw/master/img/image-20210907192331372.png)

第一我想问，上面的`split_ratio`是数据集划分吧？但再上面输出只有`train和Evaluate` 两部分，这是为什么？

第二，我想问上面的评价指标`AUC` 的值是那部分数据集的。

第三，我想问一下`Evaluate`是测试集吗，如果不是，测试集是用来最后测试的吗？

第四，如下图，这是FM训练的结束部分，我想问一下最后的`best valid`是最后的`Evaluate`的训练结果吗？

> 从这里来看好像可以说明`Evaluate` 是测试集，但如果这样，数据集划分的第二部分，验证集也就不清楚在哪里了？

![image-20210907192902968](https://gitee.com/MGzhou/zhoupic/raw/master/img/image-20210907192902968.png)

## 2 针对KGCN的问题

训练代码如下（默认模式）：

```python
from recbole.quick_start import run_recbole
run_recbole(model='KGCN', dataset='ml-100k')
```

他的其中一个配置是`split_ratio = [0.8, 0.1, 0.1]`

`eval_batch_size = 4096`

`train_batch_size = 2048`

先看我的第一个批次的输出

![image-20210907193801913](https://gitee.com/MGzhou/zhoupic/raw/master/img/image-20210907193801913.png)

我想问的问题：

我想问的是为什么`KGCN`在批次和数据集和`FM` 是一样的，但`Evaluate` 的大小上不一样，而且相差这么大？ `FM` 是`3`，而`KGCN `竟然是`472`

总共5个问题，期待你们的解答。

**Comment by hyp1231** [comment]
RecBole 默认是每个 Epoch 训练后，在验证集上评测效果。最后加载验证集上效果最好的模型参数，在测试集上只进行**一次**测试。这部分可以参考官网的介绍（如下），也可以根据 valid result / test result 进行区分。

> Note that using the quick start pipeline we provide, the original dataset will be divided into training set, validation set and test set by default. We optimize model parameters on the training set, do parameter selection according to the results on the validation set, and finally report the results on the test set.

之后逐点回复一下 5 个问题：

Q1: 输出只有 train 和 Evaluate 两部分，这是为什么？
A1: 因为 Recbole 训练结束前不在 test 集合上评测，训练过程中的 Evaluate 过程代表 valid 集上评测，下面的输出 `valid result` 也可以看出。

Q2: 上面的评价指标AUC 的值是那部分数据集的？
A2: 验证集。

Q3: Evaluate 是测试集吗，如果不是，测试集是用来最后测试的吗？
A3: Evaluate 代表一种“过程”，在训练结束前只在验证集上 Evaluate，在训练结束后只在测试集上 Evaluate；测试集是用来最后测试。

Q4: 最后的 best valid 是最后的 Evaluate 的训练结果吗？
A4: best valid 是根据 early stop 规则停止训练前，所有的 valid 结果中，最好的那一组 valid 结果；因此不是最后 Evaluate 的结果，最后 Evaluate 的结果是测试集上的。

Q5: 为什么 FM 和 KGCN 的 Evaluate 的 batch size 相差太大？
A5: 因为 FM 是 Context-Aware 类的模型，它默认是对交互记录对应的 label 做预测，因此相当于你的测试集是 10% 的交互记录，因此 3 * 4096 应该大概是 10% 交互记录的数量；而 KGCN 是 knowledge-based 类模型，继承自 general 类模型，其评测方式默认是全排序，也就是对每个 user 和每个 item 打分，因此 472 * 4096 应该大概是 10% * user 数量 * item 数量；这里建议对于 general 类 / knowledge-based 类模型，可以适当调大 eval_batch_size 若干数量级，比如 100 倍或者 1000 倍以更好地并行计算。

**Comment by MGzhou** [comment]
非常感谢你的回答！！！

**Comment by MGzhou** [comment]
我的疑问完美解决了。谢谢啦！
